From 0a7e4452bbeacb3716940060ec41fc49495327dd Mon Sep 17 00:00:00 2001
From: Subbaraya Sundeep <sbhatta@marvell.com>
Date: Wed, 30 Dec 2020 19:07:17 +0530
Subject: [PATCH 687/767] octeontx2-af: Disable packet TX and RX

commit ba9834332a3750783e1681d272f79128242e0b70 from
git@git.assembla.com:cavium/WindRiver.linux.git

The NPC RX miss stats counter increases continuously
though no interface is brought up by user. This
is because firmware enables packet I/O when
LINK_BRING_UP command is sent to it by kernel.
Hence disable packet I/O in kernel after link
up sequence. There is still a chance of some
packets reaching NPC because of the time gap
between firmware enabling packet I/O and AF
driver disabling it. This patch also disables
packet I/O during driver init for a sane
state. When interfaces are brought up then
packet I/O is enabled eventually.

Change-Id: I816c3e4631ba1bc62541ab94547a374cf936bf52
Signed-off-by: Subbaraya Sundeep <sbhatta@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/43371
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/af/cgx.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/cgx.c b/drivers/net/ethernet/marvell/octeontx2/af/cgx.c
index a579ae744b37..bdadf6ae896a 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/cgx.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/cgx.c
@@ -1398,6 +1398,10 @@ static void cgx_lmac_linkup_work(struct work_struct *work)
 		if (err)
 			dev_info(dev, "cgx port %d:%d Link up command failed\n",
 				 cgx->cgx_id, i);
+		/* Disable packet I/O in CGX in case firmware enabled it
+		 * after Link up sequence.
+		 */
+		cgx_lmac_rx_tx_enable(cgx, i, false);
 	}
 }
 
@@ -1499,6 +1503,8 @@ static int cgx_lmac_init(struct cgx *cgx)
 		/* Add reference */
 		cgx->lmac_idmap[i] = lmac;
 		cgx_lmac_pause_frm_config(cgx, i, true);
+		/* Disable packet I/O for a sane state */
+		cgx_lmac_rx_tx_enable(cgx, lmac->lmac_id, false);
 	}
 
 	return cgx_lmac_verify_fwi_version(cgx);
-- 
2.31.1

